routes/api.js

```diff
@@ -2,17 +2,81 @@ const express = require( "express" );
 const router = express.Router();
 const bcrypt = require( "bcrypt" );
 const passport = require( "passport" );
+const methodOverride = require( "method-override" );
+
+const { checkNotAuthenticated, checkAuthenticated, returnAuthentication } = require( "../util/passportUtil" );
 
-const passport_config = require( "../util/passportUtil" );
-const { checkNotAuthenticated } = passport_config;
 const randomColor = require( "../util/randomColor" );
 
+router.use( returnAuthentication );
+router.use( methodOverride( "_method" ) );
+
 router.post( "/login", passport.authenticate( "local", {
-  successRedirect: "/home",
+  successRedirect: "/app",
   failureRedirect: "/login",
   failureFlash   : true,
 } ) );
 
+router.put( "/user/password", checkAuthenticated, async ( req, res, next ) => {
+  const { password } = req.body;
+  if ( !password || password === "" )
+    return next( new Error( "400: empty user password" ) );
+
+
+  const hashedPassword = await bcrypt.hash( password, 10 );
+  const updateObj = { password: hashedPassword };
+
+  req.models.User.update( updateObj,
+    {
+      where: { id: res.locals.userid },
+    } )
+    .then( () => res.json( { error: false } ) )
+    .catch( err => next( err ) );
+  req.flash( "error", "Please sign in with new password" );
+  req.logOut();
+  res.redirect( "../../login" );
+} );
+
+router.put( "/user/username", checkAuthenticated, async ( req, res, next ) => {
+  const { username } = req.body;
+  if ( !username || username === "" )
+    return next( new Error( "400: empty username" ) );
+
+  const user = await req.models.User.findOne(
+    {
+      where: { username: req.body.username },
+      raw  : true,
+    }
+  );
+  if ( user == null ) {
+    req.models.User.update( { username }, { where: { id: res.locals.userid } } )
+      .then( () => res.json( { error: false } ) )
```

Comment:
`res.json()` schickt dem Client eine Antwort.
`res.redirect()` leitet den Client auf eine andere Seite weiter.

Es geht nur eins von beiden.

`res.redirect()` funktioniert hier nicht, da das ja eine api request (mit axios/fetch) und nicht mit dem Browser selbst.

flash & logOut also vor `res.json` in den `then` Block.

Da funktioniert auch so, weil die Datenbank zu lange braucht, also das `res.json` nicht ausgeführt wird, wenn du aber ein await vor den db call machst ist das kaputt.

Reply:
Das ist auch der Grund für den Fehler den ich dir geschickt hatte:

```500 PUT /api/user/username?_method=PUT - Cannot set headers after they are sent to the client                                                          
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client                                                                    
    at ServerResponse.setHeader (_http_outgoing.js:558:11)                                                                                             
    at ServerResponse.header (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:771:10)
    at ServerResponse.send (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:170:12)
    at done (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:1008:10)
    at Object.exports.renderFile (/home/jneidel/deployments/wd/node_modules/pug/lib/index.js:446:12)
    at View.exports.__express [as engine] (/home/jneidel/deployments/wd/node_modules/pug/lib/index.js:491:11)
    at View.render (/home/jneidel/deployments/wd/node_modules/express/lib/view.js:135:8)
    at tryRender (/home/jneidel/deployments/wd/node_modules/express/lib/application.js:640:10)
    at Function.render (/home/jneidel/deployments/wd/node_modules/express/lib/application.js:592:3)
    at ServerResponse.render (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:1012:7)
500 PUT /api/user/password?_method=PUT - Cannot set headers after they are sent to the client
Error [ERR_HTTP_HEADERS_SENT]: Cannot set headers after they are sent to the client
    at ServerResponse.setHeader (_http_outgoing.js:558:11)
    at ServerResponse.header (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:771:10)
    at ServerResponse.send (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:170:12)
    at done (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:1008:10)
    at Object.exports.renderFile (/home/jneidel/deployments/wd/node_modules/pug/lib/index.js:446:12)
    at View.exports.__express [as engine] (/home/jneidel/deployments/wd/node_modules/pug/lib/index.js:491:11)
    at View.render (/home/jneidel/deployments/wd/node_modules/express/lib/view.js:135:8)
    at tryRender (/home/jneidel/deployments/wd/node_modules/express/lib/application.js:640:10)
    at Function.render (/home/jneidel/deployments/wd/node_modules/express/lib/application.js:592:3)
    at ServerResponse.render (/home/jneidel/deployments/wd/node_modules/express/lib/response.js:1012:7)
```

Reply:
Leider ist auch nach dem fixen dieses Fehlers erneut einer aufgetreten. Local laufen jedoch weiterhin alle tests durch.